RFC: Add check-origin method for CSRF

3noch opened the pull request:
Fixes #53 

This got me thinking: the CSRF protection could probably be entirely "pluggable". When it comes to security, you don't necessarily want that much flexibility though.

This doesn't pass tests yet but I wanted to see what you thought of the approach. I don't like how it sets an empty cookie just to get two set-cookie headers to make the types happy.

3noch commented on servant-auth-server/src/Servant/Auth/Server/Internal/ConfigTypes.hs:
```diff
@@ -79,7 +75,31 @@ defaultCookieSettings = CookieSettings
+  , xsrfCookiePath :: Maybe BS.ByteString
+  -- | What name to use for the header used for CSRF protection.
+  , xsrfHeaderName :: BS.ByteString
+  } deriving (Eq, Show, Generic)
```
Should this have all the same settings as the session cookie? I think the suggestion in the README to have JS pull out the XSRF token from the cookies would not work if you use HTTP-only cookies. However, you really want to use HTTP-only cookies for the session cookie. Therefore having them the same puts you in a bind.

3noch commented:
Another possibility would be to capture the CSRF prevention technique in the API spec. Maybe that's overkill, but it would certainly be useful to add docs for your CSRF prevention. For example, something complex like this could move all the choices of cookie name and header name to the spec itself:

```haskell
type API = Auth '[Cookie "JWT-TOKEN"] '[CheckOrigin, CsrfCookie "CSRF-TOKEN" "X-CSRF-TOKEN"]
```

where `CheckOrigin` would enable a check on the `Origin` header for CSRF purposes.

jkarni commented:
Sorry for the delay, I've been away from the internet.

OWASP [suggests](https://www.owasp.org/index.php/Cross-Site_Request_Forgery_(CSRF)_Prevention_Cheat_Sheet) (a) combining CSRF token checks with origin checks and (b) using both `Origin` and `Referer` headers for the origin check. Which would suggest *adding to* rather than *changing* the current system. On the other hand, I like having at least the option of needing no JS at all, as with your approach. And I also like the idea of not having to specify a URL (as is currently the case) so that there's no annoyance with prod vs dev environments etc . So a configurable check seems reasonable. 

Just changing to `Origin` I think is problematic, though - it simply trades off the convenience of no-JS for no-domain-configuration, and (without a second, `Referer`, check) doesn't actually support all browsers.

3noch commented:
@jkarni Thanks very much for your thorough response. I also noticed the same thing after a more careful reading of the OWASP recommendations.

Do you have thoughts about how to move forward? I'd like to tweak my PR and make it something that would be suitable to be adopted into the library as a whole.

jkarni commented on servant-auth-server/src/Servant/Auth/Server/Internal/ConfigTypes.hs:
```diff
@@ -79,7 +75,31 @@ defaultCookieSettings = CookieSettings
+    , xsrfSetting       = XsrfCookie def
+    }
+
+data XsrfSetting = XsrfCookie XsrfCookieSettings
```
I would name this `XsrfCookieMatchesHeader`.

jkarni replied:
Can you add some documentation to this datatype and it's constructors? I'd link to the [OWASP page](https://www.owasp.org/index.php/Cross-Site_Request_Forgery_(CSRF)_Prevention_Cheat_Sheet).

jkarni commented on servant-auth-server/src/Servant/Auth/Server/Internal/ConfigTypes.hs:
```diff
@@ -79,7 +75,31 @@ defaultCookieSettings = CookieSettings
+
+data XsrfSetting = XsrfCookie XsrfCookieSettings
+                 | XsrfCheckOrigin (Maybe BS.ByteString -> Bool)
+                 deriving (Generic)
```
Maybe adding two constructors, `XsrfOr XsrfSetting XsrfSetting` and `XsrfAnd XsrfSetting XsrfSetting`, so that we can combine checks? 

Also, `XsrfCheckReferer`, analogous to `XsrfCheckOrigin`, would be nice to add since it's simple. Additional ones are more work and we can add as needed.

3noch replied:
I was thinking about `XsrfOr` and `XsrfAnd` as well. At that point I wonder if it makes more sense to just use a function of type `Request -> Bool`. You get more power that way without needing to rely on a custom expression DSL.

jkarni commented:
I think encoding it in the types is something we shouldn't do at all for now, until we have a clearer idea of the use-cases. In one way that's not ideal, since we're introducing variability that's not captured by the types that might be relevant to e.g. client libraries. But those libraries don't yet exist, and backwards and forward compatibility in types is much more important than elsewhere.

jkarni commented:
Could you also update the changelog?

jkarni commented:
Could you also update the changelog?

3noch commented:
Thanks for the review! I'm out of town for now but I'll get to this ASAP.

3noch commented:
@jkarni I'm planning to revisit this very soon. I've been sidetracked by other projects.

However, I realized that without expressing CSRF in the types at all it won't be possible to derive a different client API for APIs that need CSRF and those that don't. Since not all Haskell client implementations allow you to change the request before it gets sent, I'm not sure how you would add this header. You'd need to have a different derivation for clients that need the CSRF and those that don't.

3noch commented:
At the very least I think it's worth extending the `Auth` combinator to capture that information.

3noch commented:
@jkarni I'm going to punt on the discussion about what to capture in the types. I think there's something there but it's a big change and needs more thought than I can afford for now.

I've just refactored this PR to do a few things:

  * Added an arbitrary check to the cookie settings. This makes checking `Origin` and `Referer` very easy and I added a helper to do it.
  * Made XSRF cookie checks optional.
  * Since we have to break the API anyway for these changes, I also made the config options have more consistent naming ("CSRF" and "XSRF" were both in use and the record fields did not always have the same prefix which made it hard to predict them).
  * Added a `clearSession` helper to remove the session cookies. This enables logging *out*. :)

3noch commented on servant-auth-server/src/Servant/Auth/Server/Internal/Cookie.hs:
```diff
@@ -25,57 +27,96 @@ cookieAuthCheck :: FromJWT usr => CookieSettings -> JWTSettings -> AuthCheck usr
-        NotSecure -> False
-    }
+makeCsrfCookie = makeXsrfCookie
+{-# DEPRECATED makeCsrfCookie "Use makeXsrfCookie instead" #-}
```
Since `CookieSettings` changed in non-backward-compatible ways, this deprecation is only useful for people who just use the default settings.

3noch commented:
@jkarni Would you be open to merging this if I got tests passing?

mchaver commented:
@3noch Yes, if you can get the tests passing and update the version numbers and changelogs I can get it merged.

3noch commented:
@mchaver Great! Thanks.

jkarni commented:
Sorry for how long that took. Thanks a lot!

3noch commented:
@jkarni I never did get around to updating the tests!

jkarni commented:
Yeah, I only noticed that after merging. Could you send a PR with changes
to the tests?

On Tue, Dec 12, 2017 at 2:08 AM, Elliot Cameron <notifications@github.com>
wrote:

> @jkarni <https://github.com/jkarni> I never did get around to updating
> the tests!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/haskell-servant/servant-auth/pull/54#issuecomment-351005123>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABlKmjPDfuVXzp6blL0u-OuS-ZdlUFFVks5s_lCRgaJpZM4OuMvK>
> .
>

3noch commented:
@jkarni I would love to but I have been drowning in work for two months and I won't be able get to this for another few weeks at the earliest. I wish I could do it sooner.
